### DISCUSSION TITLE
counting-the-leading-zeroes-in-a-binary-number/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#36] [Announcements]
# counting-the-leading-zeroes-in-a-binary-number/

In this post I describe an algorithm to count the number of leading zeroes in a ulong, study how it works, and benchmark it against alternatives.

https://andrewlock.net/counting-the-leading-zeroes-in-a-binary-number/

**Comment by BillPete** [comment]
Nice explanation of an interesting algorithm.  This reinforces, to me, the importance of looking for improvements to brute force looping methods.  One thing to report, though: in line D, the leading one in the binary word should be a zero.  The following calculations look okay, so probably just a typo.

**Comment by khellang** [comment]
Coincidentally this is the exact same algorithm used by the [software fallback](https://github.com/dotnet/runtime/blob/161464e33a1333bb5ac3fc509172a5e2ada5af04/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs#L369-L393) for [`BitOperations.LeadingZeroCount`](https://github.com/dotnet/runtime/blob/161464e33a1333bb5ac3fc509172a5e2ada5af04/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs#L239-L247). If you run with `set COMPlus_EnableHWIntrinsic=0`, the results would probably be the same 😃

**Reply by skovsende** [reply]
> For the third step, the smallest 1 run size is four (1111), so we shit right 4 bits, and OR:

I *think* you might have a small, but funny, typo here :)

**Comment by silkfire** [comment]
Why are you still using .NET Core 2.1, I thought it was EOL?

**Reply by andrewlock** [reply]
Because unfortunately, I work on a .NET library, and a lot of people are still using it, as well as .NET Framework 🤷‍♂️

**Comment by ErickJensen** [comment]
Your first code snippet is missing the `value` actual parameter, i.e., it should be `BitOperations.LeadingZeroCount(value)`.

**Comment by SolidSoils** [comment]
Why this smearing? I think these few lines are all that is needed:

```csharp
[Theory]
[InlineData(0, 64)]
[InlineData(ulong.MaxValue, 0)]
[InlineData(0x80, 56)]
[InlineData(0x8000, 48)]
public void CountLeadingBits(ulong value, int expectedBitcount)
{
    const ulong hiBit = ulong.MaxValue ^ (ulong.MaxValue >> 1);
    int x = 0;

    while ((value & hiBit) == 0UL && x < 64)
    {
        value <<= 1;
        x++;
    }

    Assert.Equal(expectedBitcount, x);
}
```

**Reply by andrewlock** [reply]
Have you benchmarked that compared to the algorithm in my post? Speed is the reason to use smearing 🙂
